src/lib/tools/repeat_finder.rs

```diff
@@ -20,33 +29,141 @@ pub struct Opts {
 
     /// The minimum number of bases for a repeat region must span to be called a microsatellite
     #[clap(short = 'm', default_value = "10", long, display_order = 3)]
-    pub min_bases: u64,
+    pub min_bases: usize,
 
     /// The maximum number of bases for a repeat region must span to be called a microsatellite
     #[clap(short = 'M', default_value = "100", long, display_order = 4)]
-    pub max_bases: u64,
+    pub max_bases: usize,
 
     /// The minimum number of repeats for a repeat region must span to be called a microsatellite
     #[clap(short = 'r', default_value = "3", long, display_order = 5)]
-    pub min_repeats: u64,
+    pub min_repeats: usize,
 
     /// The minimum k-mer length
     #[clap(short = 'l', default_value = "1", long, display_order = 6)]
-    pub min_repeat_length: u64,
+    pub min_repeat_length: usize,
 
     /// The maximum k-mer length
     #[clap(short = 'L', default_value = "5", long, display_order = 7)]
-    pub max_repeat_length: u64,
+    pub max_repeat_length: usize,
 }
 
 // Run extract
 #[allow(clippy::too_many_lines)]
 pub fn run(opts: &Opts) -> Result<(), anyhow::Error> {
-    println!("{:?}", opts);
-    // TODO
+    // Build the finders
+    let mut finders: Vec<KmerFinder> = Vec::new();
+    let mut i = opts.min_repeat_length;
+    while i <= opts.max_repeat_length {
+        finders.push(KmerFinder::new(i));
+        i += 1;
+    }
+
+    // Open the input and output
+    let mut reader = File::open(&opts.input).map(BufReader::new).map(fasta::Reader::new)?;
+    let mut writer = File::create(&opts.output).map(BufWriter::new).map(bed::Writer::new)?;
+
+    // Go through each contig, one at a time
+    for result in reader.records() {
```

Comment:
If I were inclined to bet I'd say the slowdown vs C++ occurs here. They parse the FASTA 1 char at a time if I'm reading their code correctly (possible I'm not reading correctly, my C++ is a bit rusty (pun not intended)), meaning they don't need to read the whole contig into memory.

Reply:
The main slowdown was how we got the ith base in the sequence, using the `get` method.  Retrieving the bases as `&[u8]` and then iterating speeds this up a lot, almost comparable.